knative/describer.go

```diff
@@ -0,0 +1,100 @@
+package knative
+
+import (
+	"fmt"
+	"github.com/boson-project/faas/client"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"knative.dev/eventing/pkg/apis/eventing/v1alpha1"
+	eventingv1client "knative.dev/eventing/pkg/client/clientset/versioned/typed/eventing/v1alpha1"
+	servingv1client "knative.dev/serving/pkg/client/clientset/versioned/typed/serving/v1alpha1"
+)
+
+type Describer struct {
+	Verbose        bool
+	namespace      string
+	servingClient  *servingv1client.ServingV1alpha1Client
+	eventingClient *eventingv1client.EventingV1alpha1Client
+	config 		   *rest.Config
+}
+
+func NewDescriber(namespace string) (describer *Describer, err error) {
+	describer = &Describer{namespace: namespace}
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	clientConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, &clientcmd.ConfigOverrides{})
+	if describer.namespace == "" {
+		namespace, _, err := clientConfig.Namespace()
+		if err == nil {
+			describer.namespace = namespace
+		}
+	}
+	config, err := clientConfig.ClientConfig()
+	if err != nil {
+		return
+	}
+	describer.servingClient, err = servingv1client.NewForConfig(config)
+	if err != nil {
+		return
+	}
+	describer.eventingClient, err = eventingv1client.NewForConfig(config)
+	if err != nil {
+		return
+	}
+	describer.config = config
+	return
+}
+
+func (describer *Describer) Describe(name string) (description client.FunctionDescription, err error) {
```

Comment:
The name which is actually used in-cluster is a "DNS subdomain" (DNS 1035 label) encoded version of the fully-qualified service name as presented to the faas developer.  I.e. `www.example.com` becomes `www-example-com`.  There are utility functions for going back-and-forth between these representations in `k8s/names.go` such that you can `k8s.ToSubdomain(displayName)` when describing, and then `k8s.FromSubdomain(serviceName)` when exporting.